### PR TITLE
Document remaining attributes and macros

### LINKED ISSUE #6444: Document all attributes and macros with examples
`rust.md#attributes` and `rust.md#macros` document some attributes and macros, but give few examples. Also there are some attributes/macros missing.

Attributes:

- [x] abi  (already in rust.md; see commit 772028acdb229d1aaa5ac9ad670b276a218469a8)
- [x] allow (http://static.rust-lang.org/doc/master/rust.html#lint-check-attributes)
- [x] bench (delegated to #7824)
- [x] cfg (delegated to #7962)
- [x] deny (http://static.rust-lang.org/doc/master/rust.html#lint-check-attributes)
- [x] deriving (http://static.rust-lang.org/doc/master/rust.html#deriving)
- [x] doc (mentioned in commit 58807b05; delegated to #7963)
- [x] forbid (http://static.rust-lang.org/doc/master/rust.html#lint-check-attributes)
- [x] inline: src/libsyntax/attr.rs (delegated to #7959)
- [x] lang (handled by #3342; done; see comment dbeea18f )
- [x] license (delegated to #7961)
- [x] link (delegated to #7960; see commit d4c40b519b080a08a37d27e015083fd250fdd197)
- [x] link_args (delegated to #7960; see commit d4c40b519b080a08a37d27e015083fd250fdd197)
- [x] link_name (already in rust.md; see commit 772028acdb229d1aaa5ac9ad670b276a218469a8)
- [x] no_implicit_prelude, see #7844
- [x] nolink (delegated to #7960)
- [x] no_mangle (see librustc/middle/trans/base.rs)
- [x] no_std (see librustc/front/std_inject.rs)
- [x] path (done; see commit d4daa78a )
- [ ] should_fail (see librustc/front/test.rs)
- [x] start
- [x] static_assert (see librustc/middle/trans/base.rs)
- [x] test (delegated to #7824)
- [x] unsafe_destructor
- [x] warn (http://static.rust-lang.org/doc/master/rust.html#lint-check-attributes)

Macros:

- [x] asm http://static.rust-lang.org/doc/master/guide-unsafe.html#inline-assembly
- [x] assert http://static.rust-lang.org/doc/master/std/macros/macro.assert.html
- [x] assert_eq http://static.rust-lang.org/doc/master/std/macros/macro.assert_eq.html
- [x] cfg http://static.rust-lang.org/doc/master/std/macros/builtin/macro.cfg.html
- [ ] cold
- [x] concat_idents http://static.rust-lang.org/doc/master/std/macros/builtin/macro.concat_idents.html
- [x] bytes http://static.rust-lang.org/doc/master/std/macros/builtin/macro.bytes.html
- [x] debug (http://static.rust-lang.org/doc/master/std/logging/index.html)
- [x] env http://static.rust-lang.org/doc/master/std/macros/builtin/macro.env.html
- [x] error (http://static.rust-lang.org/doc/master/std/logging/index.html)
- [x] file (http://static.rust-lang.org/doc/master/rust.html#syntax-extensions)
- [x] format (http://static.rust-lang.org/doc/master/std/fmt/index.html)
- [x] include (https://github.com/mozilla/rust/blob/master/doc/rust.md#syntax-extensions)
- [x] include_bin (https://github.com/mozilla/rust/blob/master/doc/rust.md#syntax-extensions)
- [x] include_str (https://github.com/mozilla/rust/blob/master/doc/rust.md#syntax-extensions)
- [x] info (http://static.rust-lang.org/doc/master/std/logging/index.html)
- [x] line http://static.rust-lang.org/doc/master/std/macros/builtin/macro.line.html
- [x] log (http://static.rust-lang.org/doc/master/std/logging/index.html)
- [x] log_syntax
- [x] macro_rules
- [x] module_path http://static.rust-lang.org/doc/master/std/macros/builtin/macro.module_path.html
- [x] option_env http://static.rust-lang.org/doc/master/std/macros/builtin/macro.option_env.html
- [x] proto (is gone?)
- [ ] quote_expr
- [ ] quote_item
- [ ] quote_pat
- [ ] quote_stmt
- [ ] quote_tokens
- [ ] quote_ty
- [x] stringify (https://github.com/mozilla/rust/blob/master/doc/rust.md#syntax-extensions)
- [x] warn (http://static.rust-lang.org/doc/master/std/logging/index.html)

### PR DESCRIPTION
**Pull request by steveklabnik**
Fixes #6444

I need some help with the macro docs here, as I couldn't figure it out. @kmcallister @eddyb  or someone else?

### CODE REVIEW

**Comment by kmcallister** [comment]
The [plugins guide](http://doc.rust-lang.org/book/plugins.html) says a little about the `quote_*` macros.  They are quasiquotes for use in writing procedural macros, and they are extremely janky.

**Comment by steveklabnik** [comment]
Should I just put that for now, then?

**Comment by huonw** [comment]
I have a feeling there's been additions to our set of macros & attributes since #6444, but I am unsure.

**Comment by steveklabnik** [comment]
That's possible. Where's the definitive list?

**Comment by steveklabnik** [comment]
I think this is good enough to ship now. If there is anything else I missed, it'll get caught in the reference work.

File: src/doc/reference.md [hunk 1]
```diff
@@ -2041,6 +2050,8 @@ type int8_t = i8;
   item](#language-items) for more details.
 - `test` - indicates that this function is a test function, to only be compiled
   in case of `--test`.
+- `should_fail` - indicates that this test function should panic, inverting the success condition.
+- `cold` - The function is unlikely to be executed, so optimize it differently.
```
Comment (huonw) [inline src/doc/reference.md]: "so optimize it and calls to it", maybe? (IME, the main use is making the compiler rearrange branches to preference the ones that don't have `#[cold]` functions.)

**Comment by huonw** [comment]
(r=me)

File: src/doc/reference.md (hunk 1 above)
Comment (steveklabnik) [inline src/doc/reference.md]: seems good

**Pull request merged**